File: Leetcode/Algorithm/python/1000/00110-Balanced Binary Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isBalanced(self, root):
        (depth, balanced) = self._isBalanced(root)
        return balanced
        
    def _isBalanced(self, root):
        if not root:
            return (0, True)
        (left_depth, left_balanced)   = self._isBalanced(root.left)
        (right_depth, right_balanced) = self._isBalanced(root.right)
        depth = max(left_depth, right_depth) + 1
        balanced = abs(left_depth - right_depth) <= 1 and left_balanced and right_balanced
        return (depth, balanced)

File: Leetcode/Algorithm/cpp/0000/00110-Balanced Binary Tree.cc
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        res = true;
        if (!root) {
            return true;
        }
        do_is_balanced(root);
        return res;
    }
    
    int do_is_balanced(TreeNode *root) {
        if (!root) {
            return 0;
        }
        int l = do_is_balanced(root -> left);
        int r = do_is_balanced(root -> right);
        
        if (abs(l - r) > 1) {
            res = false;
        }
        return max(l, r) + 1;
    }
private:
    bool res;
};
